File: apps/api/app/services/plaid_service.py
"""
Tally API — Plaid service.
Wraps the plaid-python SDK for Link, account fetch, and transaction sync.
"""

from datetime import UTC, date as date_type, datetime

from plaid.api import plaid_api
from plaid.api_client import ApiClient
from plaid.configuration import Configuration
from plaid.model.accounts_get_request import AccountsGetRequest
from plaid.model.country_code import CountryCode
from plaid.model.item_public_token_exchange_request import ItemPublicTokenExchangeRequest
from plaid.model.item_remove_request import ItemRemoveRequest
from plaid.model.link_token_create_request import LinkTokenCreateRequest
from plaid.model.link_token_create_request_user import LinkTokenCreateRequestUser
from plaid.model.products import Products
from plaid.model.transactions_sync_request import TransactionsSyncRequest
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.config import get_settings
from app.core.encryption import decrypt_token, encrypt_token
from app.models.account import Account
from app.models.plaid_item import PlaidItem
from app.models.transaction import Transaction
from app.models.user import User

settings = get_settings()

PLAID_ENV_MAP = {
    "sandbox": "https://sandbox.plaid.com",
    "development": "https://development.plaid.com",
    "production": "https://production.plaid.com",
}


def _as_dict(obj: object) -> dict:
    if isinstance(obj, dict):
        return obj
    if hasattr(obj, "to_dict"):
        return obj.to_dict()
    return dict(obj)


def _get_plaid_client() -> plaid_api.PlaidApi:
    configuration = Configuration(
        host=PLAID_ENV_MAP[settings.plaid_env],
        api_key={
            "clientId": settings.plaid_client_id,
            "secret": settings.plaid_secret,
        },
    )
    return plaid_api.PlaidApi(ApiClient(configuration))


def create_link_token(user_id: str) -> tuple[str, str]:
    client = _get_plaid_client()
    request = LinkTokenCreateRequest(
        user=LinkTokenCreateRequestUser(client_user_id=user_id),
        client_name="Tally",
        products=[Products("transactions")],
        country_codes=[CountryCode("US")],
        language="en",
    )
    response = client.link_token_create(request)
    data = _as_dict(response)
    expiration = data["expiration"]
    exp_str = expiration.isoformat() if hasattr(expiration, "isoformat") else str(expiration)
    return data["link_token"], exp_str


async def exchange_public_token(
    db: AsyncSession,
    user: User,
    public_token: str,
    institution_id: str | None,
    institution_name: str | None,
) -> PlaidItem:
    client = _get_plaid_client()
    exchange_request = ItemPublicTokenExchangeRequest(public_token=public_token)
    exchange_response = client.item_public_token_exchange(exchange_request)
    exchange_data = _as_dict(exchange_response)
    access_token = exchange_data["access_token"]
    plaid_item_id = exchange_data["item_id"]

    item = PlaidItem(
        user_id=user.id,
        plaid_item_id=plaid_item_id,
        access_token_encrypted=encrypt_token(access_token),
        institution_id=institution_id,
        institution_name=institution_name,
    )
    db.add(item)
    await db.flush()

    await _upsert_accounts(db, client, item, access_token, institution_name)
    await sync_transactions_for_item(db, item)
    return item


async def _upsert_accounts(
    db: AsyncSession,
    client: plaid_api.PlaidApi,
    item: PlaidItem,
    access_token: str,
    institution_name: str | None,
) -> list[Account]:
    response = client.accounts_get(AccountsGetRequest(access_token=access_token))
    response_data = _as_dict(response)
    accounts: list[Account] = []

    for acct in response_data.get("accounts", []):
        acct = _as_dict(acct)
        plaid_account_id = acct["account_id"]
        result = await db.execute(
            select(Account).where(Account.plaid_account_id == plaid_account_id)
        )
        existing = result.scalar_one_or_none()

        balances = _as_dict(acct.get("balances") or {})
        current = float(balances.get("current") or 0)
        available = balances.get("available")
        available_float = float(available) if available is not None else None

        if existing:
            existing.name = acct.get("name") or existing.name
            existing.official_name = acct.get("official_name")
            existing.type = acct.get("type") or existing.type
            existing.subtype = acct.get("subtype")
            existing.balance_current = current
            existing.balance_available = available_float
            existing.institution_name = institution_name
            existing.last_synced_at = datetime.now(UTC)
            accounts.append(existing)
        else:
            account = Account(
                plaid_account_id=plaid_account_id,
                user_id=item.user_id,
                plaid_item_id=item.id,
                name=acct.get("name") or "Account",
                official_name=acct.get("official_name"),
                type=acct.get("type") or "other",
                subtype=acct.get("subtype"),
                balance_current=current,
                balance_available=available_float,
                institution_name=institution_name,
                last_synced_at=datetime.now(UTC),
            )
            db.add(account)
            accounts.append(account)

    await db.flush()
    return accounts


async def sync_transactions_for_item(db: AsyncSession, item: PlaidItem) -> int:
    client = _get_plaid_client()
    access_token = decrypt_token(item.access_token_encrypted)
    cursor = item.cursor
    added_count = 0
    has_more = True

    account_map: dict[str, Account] = {}
    result = await db.execute(select(Account).where(Account.plaid_item_id == item.id))
    for account in result.scalars().all():
        account_map[account.plaid_account_id] = account

    while has_more:
        request = TransactionsSyncRequest(access_token=access_token, cursor=cursor)
        response = client.transactions_sync(request)
        sync_data = _as_dict(response)

        for plaid_tx in sync_data.get("added", []):
            account = account_map.get(_as_dict(plaid_tx)["account_id"])
            if not account:
                continue
            await _upsert_transaction(db, account, _as_dict(plaid_tx))
            added_count += 1

        for plaid_tx in sync_data.get("modified", []):
            account = account_map.get(_as_dict(plaid_tx)["account_id"])
            if not account:
                continue
            await _upsert_transaction(db, account, _as_dict(plaid_tx))

        for plaid_tx in sync_data.get("removed", []):
            removed = _as_dict(plaid_tx)
            tx_id = removed.get("transaction_id")
            if not tx_id:
                continue
            result = await db.execute(
                select(Transaction).where(Transaction.plaid_transaction_id == tx_id)
            )
            existing = result.scalar_one_or_none()
            if existing:
                await db.delete(existing)

        cursor = sync_data.get("next_cursor")
        has_more = sync_data.get("has_more", False)

    item.cursor = cursor
    item.last_synced_at = datetime.now(UTC)
    await db.flush()
    return added_count


async def _upsert_transaction(
    db: AsyncSession, account: Account, plaid_tx: dict
) -> Transaction:
    plaid_transaction_id = plaid_tx["transaction_id"]
    result = await db.execute(
        select(Transaction).where(Transaction.plaid_transaction_id == plaid_transaction_id)
    )
    existing = result.scalar_one_or_none()

    amount = float(plaid_tx.get("amount") or 0)
    tx_type = "debit" if amount > 0 else "credit"

    pfc = _as_dict(plaid_tx.get("personal_finance_category") or {})
    category = pfc.get("primary") or (
        (plaid_tx.get("category") or [None])[0] if plaid_tx.get("category") else None
    )

    tx_date = plaid_tx["date"]
    if isinstance(tx_date, str):
        tx_date = date_type.fromisoformat(tx_date)

    fields = {
        "account_id": account.id,
        "amount": amount,
        "type": tx_type,
        "name": plaid_tx.get("name") or "Transaction",
        "merchant_name": plaid_tx.get("merchant_name") or plaid_tx.get("name"),
        "category": category,
        "date": tx_date,
        "pending": bool(plaid_tx.get("pending", False)),
        "logo_url": plaid_tx.get("logo_url"),
    }

    if existing:
        for key, value in fields.items():
            setattr(existing, key, value)
        return existing

    tx = Transaction(plaid_transaction_id=plaid_transaction_id, **fields)
    db.add(tx)
    return tx


async def sync_all_user_items(db: AsyncSession, user_id: str) -> int:
    result = await db.execute(select(PlaidItem).where(PlaidItem.user_id == user_id))
    items = result.scalars().all()
    total = 0
    for item in items:
        client = _get_plaid_client()
        access_token = decrypt_token(item.access_token_encrypted)
        await _upsert_accounts(db, client, item, access_token, item.institution_name)
        total += await sync_transactions_for_item(db, item)
    return total


async def remove_plaid_item(db: AsyncSession, user_id: str, item_id: str) -> None:
    result = await db.execute(
        select(PlaidItem).where(PlaidItem.id == item_id, PlaidItem.user_id == user_id)
    )
    item = result.scalar_one_or_none()
    if not item:
        return

    client = _get_plaid_client()
    access_token = decrypt_token(item.access_token_encrypted)
    try:
        client.item_remove(ItemRemoveRequest(access_token=access_token))
    except Exception:
        pass  # Item may already be removed on Plaid side

    await db.delete(item)

File: apps/api/src/main/java/com/tally/finance/service/PlaidService.java
package com.tally.finance.service;

import com.plaid.client.ApiClient;
import com.plaid.client.model.*;
import com.plaid.client.request.PlaidApi;
import com.tally.finance.config.TallyProperties;
import com.tally.finance.domain.Account;
import com.tally.finance.domain.PlaidItem;
import com.tally.finance.domain.Transaction;
import com.tally.finance.domain.User;
import com.tally.finance.repository.AccountRepository;
import com.tally.finance.repository.PlaidItemRepository;
import com.tally.finance.repository.TransactionRepository;
import java.io.IOException;
import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class PlaidService {

    private final TallyProperties properties;
    private final TokenEncryptionService encryptionService;
    private final PlaidItemRepository plaidItemRepository;
    private final AccountRepository accountRepository;
    private final TransactionRepository transactionRepository;

    public PlaidService(
            TallyProperties properties,
            TokenEncryptionService encryptionService,
            PlaidItemRepository plaidItemRepository,
            AccountRepository accountRepository,
            TransactionRepository transactionRepository
    ) {
        this.properties = properties;
        this.encryptionService = encryptionService;
        this.plaidItemRepository = plaidItemRepository;
        this.accountRepository = accountRepository;
        this.transactionRepository = transactionRepository;
    }

    public LinkTokenResult createLinkToken(String userId) throws IOException {
        PlaidApi client = plaidClient();
        LinkTokenCreateRequest request = new LinkTokenCreateRequest()
                .user(new LinkTokenCreateRequestUser().clientUserId(userId))
                .clientName("Tally")
                .products(List.of(Products.TRANSACTIONS))
                .countryCodes(List.of(CountryCode.US))
                .language("en");

        LinkTokenCreateResponse response = client.linkTokenCreate(request).execute().body();
        if (response == null) {
            throw new IOException("Empty Plaid link token response");
        }
        return new LinkTokenResult(response.getLinkToken(), response.getExpiration().toString());
    }

    @Transactional
    public PlaidItem exchangePublicToken(
            User user,
            String publicToken,
            String institutionId,
            String institutionName
    ) throws IOException {
        PlaidApi client = plaidClient();
        ItemPublicTokenExchangeRequest exchangeRequest =
                new ItemPublicTokenExchangeRequest().publicToken(publicToken);
        ItemPublicTokenExchangeResponse exchangeResponse =
                client.itemPublicTokenExchange(exchangeRequest).execute().body();
        if (exchangeResponse == null) {
            throw new IOException("Empty Plaid exchange response");
        }

        String accessToken = exchangeResponse.getAccessToken();
        String plaidItemId = exchangeResponse.getItemId();

        PlaidItem item = new PlaidItem();
        item.setUser(user);
        item.setPlaidItemId(plaidItemId);
        item.setAccessTokenEncrypted(encryptionService.encrypt(accessToken));
        item.setInstitutionId(institutionId);
        item.setInstitutionName(institutionName);
        item = plaidItemRepository.save(item);

        upsertAccounts(client, item, accessToken, institutionName);
        syncTransactionsForItem(item);
        return item;
    }

    @Transactional
    public int syncAllUserItems(String userId) throws IOException {
        List<PlaidItem> items = plaidItemRepository.findByUser_Id(userId);
        int total = 0;
        for (PlaidItem item : items) {
            // Incremental sync only — uses stored cursor; does not re-pull full history.
            total += syncTransactionsForItem(item);
        }
        return total;
    }

    @Transactional
    public void removePlaidItem(String userId, String itemId) throws IOException {
        PlaidItem item = plaidItemRepository.findByIdAndUser_Id(itemId, userId).orElse(null);
        if (item == null) {
            return;
        }

        try {
            String accessToken = encryptionService.decrypt(item.getAccessTokenEncrypted());
            plaidClient().itemRemove(new ItemRemoveRequest().accessToken(accessToken)).execute();
        } catch (Exception ignored) {
            // Item may already be removed on Plaid side
        }

        plaidItemRepository.delete(item);
    }

    @Transactional
    public int syncTransactionsForItem(PlaidItem item) throws IOException {
        PlaidApi client = plaidClient();
        String accessToken = encryptionService.decrypt(item.getAccessTokenEncrypted());
        String cursor = item.getCursor();
        int addedCount = 0;
        boolean hasMore = true;

        Map<String, Account> accountMap = new HashMap<>();
        for (Account account : accountRepository.findByPlaidItemId(item.getId())) {
            accountMap.put(account.getPlaidAccountId(), account);
        }

        while (hasMore) {
            TransactionsSyncRequest request = new TransactionsSyncRequest()
                    .accessToken(accessToken)
                    .cursor(cursor);
            TransactionsSyncResponse syncData = client.transactionsSync(request).execute().body();
            if (syncData == null) {
                break;
            }

            for (com.plaid.client.model.Transaction plaidTx : syncData.getAdded()) {
                Account account = accountMap.get(plaidTx.getAccountId());
                if (account != null) {
                    upsertTransaction(account, plaidTx);
                    addedCount++;
                }
            }

            for (com.plaid.client.model.Transaction plaidTx : syncData.getModified()) {
                Account account = accountMap.get(plaidTx.getAccountId());
                if (account != null) {
                    upsertTransaction(account, plaidTx);
                }
            }

            for (RemovedTransaction removed : syncData.getRemoved()) {
                if (removed.getTransactionId() != null) {
                    transactionRepository.findByPlaidTransactionId(removed.getTransactionId())
                            .ifPresent(transactionRepository::delete);
                }
            }

            cursor = syncData.getNextCursor();
            hasMore = Boolean.TRUE.equals(syncData.getHasMore());
        }

        item.setCursor(cursor);
        item.setLastSyncedAt(Instant.now());
        plaidItemRepository.save(item);
        return addedCount;
    }

    private void upsertAccounts(
            PlaidApi client,
            PlaidItem item,
            String accessToken,
            String institutionName
    ) throws IOException {
        AccountsGetResponse response =
                client.accountsGet(new AccountsGetRequest().accessToken(accessToken)).execute().body();
        if (response == null) {
            return;
        }

        Instant now = Instant.now();
        for (AccountBase acct : response.getAccounts()) {
            var existing = accountRepository.findByPlaidAccountId(acct.getAccountId());
            Account account = existing.orElseGet(Account::new);

            AccountBalance balances = acct.getBalances();
            BigDecimal current = balances.getCurrent() != null
                    ? BigDecimal.valueOf(balances.getCurrent())
                    : BigDecimal.ZERO;
            BigDecimal available = balances.getAvailable() != null
                    ? BigDecimal.valueOf(balances.getAvailable())
                    : null;

            if (existing.isEmpty()) {
                account.setPlaidAccountId(acct.getAccountId());
                account.setUser(item.getUser());
                account.setPlaidItem(item);
                account.setName(Objects.toString(acct.getName(), "Account"));
            } else {
                account.setName(Objects.toString(acct.getName(), account.getName()));
            }

            account.setOfficialName(acct.getOfficialName());
            account.setType(acct.getType() != null ? acct.getType().getValue() : "other");
            account.setSubtype(acct.getSubtype() != null ? String.valueOf(acct.getSubtype()) : null);
            account.setBalanceCurrent(current);
            account.setBalanceAvailable(available);
            account.setInstitutionName(institutionName);
            account.setLastSyncedAt(now);
            accountRepository.save(account);
        }
    }

    private void upsertTransaction(Account account, com.plaid.client.model.Transaction plaidTx) {
        String plaidTransactionId = plaidTx.getTransactionId();
        var existing = transactionRepository.findByPlaidTransactionId(plaidTransactionId);
        Transaction tx = existing.orElseGet(Transaction::new);

        double amountValue = plaidTx.getAmount() != null ? plaidTx.getAmount() : 0.0;
        String txType = amountValue > 0 ? "debit" : "credit";

        String category = null;
        if (plaidTx.getPersonalFinanceCategory() != null
                && plaidTx.getPersonalFinanceCategory().getPrimary() != null) {
            category = plaidTx.getPersonalFinanceCategory().getPrimary();
        } else if (plaidTx.getCategory() != null && !plaidTx.getCategory().isEmpty()) {
            category = plaidTx.getCategory().get(0);
        }

        LocalDate txDate = plaidTx.getDate();
        if (txDate == null && plaidTx.getAuthorizedDate() != null) {
            txDate = plaidTx.getAuthorizedDate();
        }

        if (existing.isEmpty()) {
            tx.setPlaidTransactionId(plaidTransactionId);
        }
        tx.setAccount(account);
        tx.setAmount(BigDecimal.valueOf(amountValue));
        tx.setType(txType);
        tx.setName(Objects.toString(plaidTx.getName(), "Transaction"));
        tx.setMerchantName(plaidTx.getMerchantName() != null ? plaidTx.getMerchantName() : tx.getName());
        tx.setCategory(category);
        tx.setDate(txDate);
        tx.setPending(Boolean.TRUE.equals(plaidTx.getPending()));
        tx.setLogoUrl(plaidTx.getLogoUrl());
        transactionRepository.save(tx);
    }

    private PlaidApi plaidClient() {
        Map<String, String> apiKeys = new HashMap<>();
        apiKeys.put("clientId", properties.plaid().clientId());
        apiKeys.put("secret", properties.plaid().secret());

        ApiClient apiClient = new ApiClient(apiKeys);
        apiClient.setPlaidAdapter(switch (properties.plaid().env()) {
            case "production" -> ApiClient.Production;
            default -> ApiClient.Sandbox;
        });
        return apiClient.createService(PlaidApi.class);
    }

    public boolean isConfigured() {
        return properties.plaid().clientId() != null && !properties.plaid().clientId().isBlank()
                && properties.plaid().secret() != null && !properties.plaid().secret().isBlank();
    }

    public record LinkTokenResult(String linkToken, String expiration) {}
}
